### PR TITLE
feat: add code coverage CI

### LINKED ISSUE #14255: Calculate code coverage
### Description

Code coverage is a great asset to have as now we don't really know which parts of the code base are tested and not.

It is possible to get test coverage in rust, running python tests with pyo3. You would need to build polars with the code coverage instrumentation flag '-C instrument-coverage', see [rustc](https://doc.rust-lang.org/rustc/instrument-coverage.html).

Py03 has an [CI workflow](https://github.com/PyO3/pyo3/blob/main/.github/workflows/ci.yml#L352) running python & rust tests to get full coverage and [pydantic-core](https://github.com/pydantic/pydantic-core) has a similar setup with python tests and rust code ([Github Action](https://github.com/pydantic/pydantic-core/blob/main/.github/workflows/ci.yml)). There is also an example project which runs a showcase [rust-python-coverage](https://github.com/cjermain/rust-python-coverage). Steps below are taken from the readme of that page

TLDR: We should be able to get code coverage running our python tests

# Step 1: Setup: Install LLVM instrumentation tooling

```
rustup component add llvm-tools-preview
cargo install cargo-llvm-cov
```

# Step 2: Set Env vars

You can easily set all the environment variables for rustc by running the following commands

```
source <(cargo llvm-cov show-env --export-prefix)
export CARGO_TARGET_DIR=$CARGO_LLVM_COV_TARGET_DIR
export CARGO_INCREMENTAL=1
```

# Step 3: Compile Polars with LLVM instrumentation

I am assuming here you have a venv installed with all the packages from `requirements-dev.txt`

```
cargo llvm-cov clean --workspace
maturin develop
pytest tests --cov=polars --cov-report xml
cargo llvm-cov --no-run --lcov --output-path coverage.lcov
```

# Segfault

Compilation runs fine with `maturin develop` and the pytest also run fine, however at the end of the run I get a segfault. Running the same command in gdb gives me the following stacktrace

I am running only a single test here, for performance reasons
```
gdb --args python3 -m coverage run -m pytest -k test_enum_creation
```

```
=========================================================== 1 passed, 5713 deselected, 121 warnings in 13.33s ===========================================================
Thread 1 "python3" received signal SIGBUS, Bus error.
WriteBinaryIds (Writer=<optimized out>, Note=0x80007d262000, NotesEnd=0x80007d262024) at ../../src/llvm-project/compiler-rt/lib/profile/InstrProfilingPlatformLinux.c:144
144	../../src/llvm-project/compiler-rt/lib/profile/InstrProfilingPlatformLinux.c: No such file or directory.
```

With a backtrace of
```
#0  WriteBinaryIds (Writer=<optimized out>, Note=0x80007d262000, NotesEnd=0x80007d262024)
    at ../../src/llvm-project/compiler-rt/lib/profile/InstrProfilingPlatformLinux.c:144
#1  __llvm_write_binary_ids (Writer=Writer@entry=0x0) at ../../src/llvm-project/compiler-rt/lib/profile/InstrProfilingPlatformLinux.c:195
#2  0x00007ffff14cc82f in lprofWriteDataImpl (Writer=Writer@entry=0x7fffffffdb38, DataBegin=DataBegin@entry=0x7ffff471d000, 
    DataEnd=DataEnd@entry=0x7ffff5646a00 <lprofValueProfNodes>, CountersBegin=CountersBegin@entry=0x7ffff41a7098 "", 
    CountersEnd=CountersEnd@entry=0x7ffff471d000 "/\230\204\060\032\342\030خ\031>\373\v\034\342Ø\240\250\377\377\377\377\377", 
    VPDataReader=0x7ffff41a7010 <TheVPDataReader>, 
    NamesBegin=0x7ffff18609e4 "\211%\333\005x\332Ֆko\233\060\024\206\265\177\204\303-|lXRѮ4\275\254\333\372\345\310\065nB\a6\263MZ\367\327\317\\\262\244S\322v\004UC d\371\362\372\071\027\037\f\227q\030\257\276ǪzBI.\217\355\300\216Nt(", NamesEnd=0x7ffff1d63a0e "", SkipNameDataWrite=0)
    at ../../src/llvm-project/compiler-rt/include/profile/InstrProfData.inc:130
#3  0x00007ffff14cc70c in lprofWriteData (Writer=Writer@entry=0x7fffffffdb38, VPDataReader=<optimized out>, SkipNameDataWrite=<optimized out>, 
    SkipNameDataWrite@entry=0) at ../../src/llvm-project/compiler-rt/lib/profile/InstrProfilingWriter.c:251
#4  0x00007ffff14caa09 in writeFile (OutputName=OutputName@entry=0x7fffffffdb80 "default_2862098967233759928_0_6947.profraw")
    at ../../src/llvm-project/compiler-rt/lib/profile/InstrProfilingFile.c:480
#5  0x00007ffff14ca782 in __llvm_profile_write_file () at ../../src/llvm-project/compiler-rt/lib/profile/InstrProfilingFile.c:1065
#6  0x00007ffff7c45495 in __run_exit_handlers (status=0, listp=0x7ffff7e1a838 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, 
    run_dtors=run_dtors@entry=true) at ./stdlib/exit.c:113
#7  0x00007ffff7c45610 in __GI_exit (status=<optimized out>) at ./stdlib/exit.c:143
#8  0x00007ffff7c29d97 in __libc_start_call_main (main=main@entry=0x55555577bff0, argc=argc@entry=8, argv=argv@entry=0x7fffffffdd58)
    at ../sysdeps/nptl/libc_start_call_main.h:74
#9  0x00007ffff7c29e40 in __libc_start_main_impl (main=0x55555577bff0, argc=8, argv=0x7fffffffdd58, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffdd48) at ../csu/libc-start.c:392
#10 0x000055555577bf25 in _start ()
```

It seems when writing the profiling information to `.profraw` it crashes.  Quick search shows: pydantic has had a similar https://github.com/pydantic/pydantic-core/issues/451, but running on Ubuntu20.04 or MacOs fixed it. I tried on Ubuntu 20.04 and 22.04, but both failed for me. 

I was hoping someone could be of help here as to how to further investigate this.

### PR DESCRIPTION
**Pull request by c-peters**
fix #14255 

I have added back in numerous integration tests from the forked arrow2 / parquet2 repositories and removed tests that were not applicable to our use of arrow. The code coverage overall is good, but in `polars-arrow` we are lower due to the fact that we don't use several datatypes of the package.

For an overview of https://app.codecov.io/gh/c-peters/polars/tree/codecov/

The CI now runs on main and in the PR with a comment on diff

### CODE REVIEW

**Comment by ritchie46** [comment]
Feel free to merge. This has a high probability of conflicts. And, really nice work!
